### PR TITLE
CLDSRV-559 AWS KMS backend + account level encryption key

### PR DESCRIPTION
**Pull request by nicolas2bert**
- Introduce a new KMS backend that supports the AWS KMS protocol.
- Implement a more efficient approach to managing default encryption keys. Instead of creating a master encryption key per bucket, a single default master encryption key is associated with the account. This simplifies key management. The default key id will be stored in the account metadata.

### CODE REVIEW

**Comment by nicolas2bert** [comment]
ping

File: tests/functional/aws-node-sdk/test/object/encryptionHeaders.js [hunk 1]
```diff
@@ -167,46 +170,51 @@ describe('per object encryption headers', () => {
...
-                    };
-                    // no op putBucketNotification for the unencrypted case
-                    const s3Op = existing.algo ? (...args) => s3.putBucketEncryption(...args) : s3NoOp;
-                    s3Op(params, error => {
-                        assert.ifError(error);
-                        return putEncryptedObject(s3, bucket, object, target, kmsKeyId, error => {
+                .forEach(existing => {
+                    const hasKey = target.masterKeyId ? 'a' : 'no';
+                    const { algo } = target;
+                    it('should override bucket encryption settings with '
```
Comment (nicolas2bert) [inline tests/functional/aws-node-sdk/test/object/encryptionHeaders.js:176]: Note to reviewer: Here I have not changed anything in the test, just improved the description a bit.

File: lib/Config.js [hunk 2]
```diff
@@ -1092,6 +1092,34 @@ class Config extends EventEmitter {
...
+        // Use env variables as default values.
+        // We use the same env variables as the AWS CLI does.
+        // Please note that if no config is specified here, the AWS Client
+        // seems to fallback on the local AWS configuration files
+        // (those contained in ~/.aws directory)
+        this.kms_aws = {
+            region: process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION,
+            endpoint: process.env.AWS_ENDPOINT_URL_KMS || process.env.AWS_ENDPOINT_URL,
+            ak: process.env.AWS_ACCESS_KEY_ID,
+            sk: process.env.AWS_SECRET_ACCESS_KEY
```
Comment (jonathan-gramain) [inline lib/Config.js]:
```suggestion
            accessKey: process.env.AWS_ACCESS_KEY_ID,
            secretKey: process.env.AWS_SECRET_ACCESS_KEY
```

File: lib/kms/wrapper.js [hunk 3]
```diff
@@ -186,27 +171,54 @@ class KMS {
...
+                    encrypt the datakey. This method is used when the KMS backend doesn't
+                    provide the generateDataKey method.
+                */
+                if (client.generateDataKey) {
+                    log.debug('creating a data key using the KMS');
+                    return client.generateDataKey(cipherBundle.cryptoScheme,
+                        cipherBundle.masterKeyId,
+                        log, (err, plainTextDataKey, cipheredDataKey) => {
+                            if (err) {
+                                log.debug('error from kms',
```
Comment (jonathan-gramain) [inline lib/kms/wrapper.js]: You may want to log this with error level? Your call, and the original code was also logging in debug level. Also maybe add some context like "error generating a new data key from KMS"

File: lib/kms/wrapper.js [hunk 4]
```diff
@@ -186,27 +171,54 @@ class KMS {
...
+                                    { implName, error: err });
+                                return next(err);
+                            }
+                            log.trace('data key generated by the kms');
+                            return next(null, plainTextDataKey, cipheredDataKey);
+                        })
+                } else {
+                    log.debug('creating a new data key');
+                    const dataKey = Common.createDataKey();
```
Comment (jonathan-gramain) [inline lib/kms/wrapper.js]:
```suggestion
                    const plainTextDataKey = Common.createDataKey();
```

File: lib/kms/wrapper.js [hunk 5]
```diff
@@ -45,43 +46,68 @@ if (config.backends.kms === 'mem') {
...
         log.debug('creating a new bucket key');
-        client.createBucketKey(bucketName, log, (err, masterKeyId) => {
+        // Check if the client supports the use of a default master encryption key per account
+        // and one is configured.
+        // If so, retrieve or create the encryption key for the account from Vault.
+        // Later its id will be stored at the bucket metadata level.
+        if (client.supportsDefaultKeyPerAccount && config.defaultEncryptionKeyPerAccount) {
+            return vault.getOrCreateEncryptionKeyId(bucket.getOwner(), log, (err, data) => {
+                if (err) {
+                    log.debug('error retreiving or creating the default encryption key at the account level from vault',
```
Comment (jonathan-gramain) [inline lib/kms/wrapper.js]:
```suggestion
                    log.debug('error retrieving or creating the default encryption key at the account level from vault',
```

File: lib/kms/wrapper.js [hunk 6]
```diff
@@ -45,43 +46,68 @@ if (config.backends.kms === 'mem') {
...
+                        { implName, error: err });
+                    return cb(err);
+                }
+
+                const { encryptionKeyId, action } = data;
+                log.trace('default encryption key retreived or created at the account level from vault',
```
Comment (jonathan-gramain) [inline lib/kms/wrapper.js]:
```suggestion
                log.trace('default encryption key retrieved or created at the account level from vault',
```

File: tests/unit/api/bucketDelete.js [hunk 7]
```diff
@@ -175,4 +180,95 @@ describe('bucketDelete API', () => {
...
+
+        beforeEach(() => {
+            destroyBucketKeySpy = sinon.spy(inMemory, 'destroyBucketKey');
+        });
+
+        afterEach(() => {
+            sinon.restore();
+        });
+
+        it('should delete the bucket-level encryption key is AES256 algorithm', done => {
```
Comment (jonathan-gramain) [inline tests/unit/api/bucketDelete.js]:
```suggestion
        it('should delete the bucket-level encryption key if AES256 algorithm', done => {
```

**Review by jonathan-gramain — APPROVED** [review]

File: lib/kms/wrapper.js [hunk 8]
```diff
@@ -181,27 +206,56 @@ class KMS {
...
-                            return next(err);
-                        }
-                        log.trace('data key ciphered by the kms');
-                        return next(null, cipheredDataKey);
-                    });
+            function generateDataKey(next) {
+                /* There are 2 ways of generating a datakey :
+                  - using the generateDataKey of the KMS backend if it exists
+                    (currently only implemented for the AWS KMS backend). This is
+                    the prefered solution since a dedicated KMS should offer a better
```
Comment (BourgoisMickael) [inline lib/kms/wrapper.js]:
```suggestion
                    the preferred solution since a dedicated KMS should offer a better
```

**Review by BourgoisMickael — APPROVED** [review]

**Comment by nicolas2bert** [comment]
@bert-e approve

**Pull request merged**